Implement more smart pointers in WebResourceLoader.h/cpp

lopezlozoya opened the pull request:
#### cac61b474e4467da3640ab13a411a21d1d7e12f6
<pre>
Implement more smart pointers in WebResourceLoader.h/cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=287211">https://bugs.webkit.org/show_bug.cgi?id=287211</a>

Reviewed by Ryosuke Niwa.

Addressed the smart pointer static analyzer warnings.

* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
* Source/WebKit/WebProcess/Network/WebResourceLoader.h:
(WebKit::WebResourceLoader::messageSenderDestinationID const):
(WebKit::WebResourceLoader::mainFrameMainResource const):
(WebKit::WebResourceLoader::willSendRequest):
(WebKit::WebResourceLoader::didSendData):
(WebKit::WebResourceLoader::didReceiveResponse):
(WebKit::WebResourceLoader::didReceiveData):
(WebKit::WebResourceLoader::didFinishResourceLoad):
(WebKit::WebResourceLoader::didFailServiceWorkerLoad):
(WebKit::WebResourceLoader::serviceWorkerDidNotHandle):
(WebKit::WebResourceLoader::didFailResourceLoad):
(WebKit::WebResourceLoader::didBlockAuthenticationChallenge):
(WebKit::WebResourceLoader::stopLoadingAfterXFrameOptionsOrContentSecurityPolicyDenied):
(WebKit::WebResourceLoader::didReceiveResource):
(WebKit::WebResourceLoader::contentFilterDidBlockLoad):

Canonical link: <a href="https://commits.webkit.org/290084@main">https://commits.webkit.org/290084@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/acfb5960e9be9538ea8dff2cf8238ee0eec1e172

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88918 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8442 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43401 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/93889 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39677 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8829 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16626 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68501 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26181 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91920 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6727 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80395 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48866 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6478 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38786 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76841 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/35715 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95729 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16098 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11744 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77379 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16354 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76266 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76667 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21054 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19479 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/9164 "The change is no longer eligible for processing.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13931 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16112 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21423 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/15853 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19304 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17634 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -87,7 +87,8 @@ IPC::Connection* WebResourceLoader::messageSenderConnection() const
     RELEASE_ASSERT(RunLoop::isMain());
-    return m_coreLoader->identifier()->toUInt64();
+    RefPtr coreLoader = m_coreLoader;
+    return coreLoader->identifier()->toUInt64();
```
We can just do:
```cpp
    return RefPtr { m_coreLoader }->identifier()->toUInt64();
```

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -115,39 +117,43 @@ MainFrameMainResource WebResourceLoader::mainFrameMainResource() const
 {
-    m_coreLoader->didSendData(bytesSent, totalBytesToBeSent);
+    RefPtr coreLoader = m_coreLoader;
+    coreLoader->didSendData(bytesSent, totalBytesToBeSent);
```
Ditto. We could do:
```cpp
RefPtr { m_coreLoader }->didSendData(bytesSent, totalBytesToBeSent);
```

But in this case, we could add a new helper like `protectedCoreLoader()` which returns RefPtr and simply use that:
```cpp
    protectedCoreLoader()->didSendData(bytesSent, totalBytesToBeSent);
```

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -166,37 +172,40 @@ void WebResourceLoader::didReceiveResponse(ResourceResponse&& response, PrivateR
                 }
 
-                RefPtr<WebCore::ResourceLoader> protectedCoreLoader = m_coreLoader;
+                RefPtr<WebCore::ResourceLoader> protectedCoreLoader = coreLoader;
```
This variable is redundant now that we have `coreLoader` which is already a `RefPtr`.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -98,15 +98,15 @@ void WebResourceLoader::detachFromCoreLoader()
 MainFrameMainResource WebResourceLoader::mainFrameMainResource() const
 {
-    RefPtr frame = m_coreLoader->frame();
+    RefPtr frame = protectedCoreLoader()->frame();
```
Oh, in this function, we probably want to declare a local variable instead of repeatedly calling protectedCoreLoader():
```
RefPtr coreLoader = m_coreLoader;
```

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -122,32 +122,32 @@ void WebResourceLoader::willSendRequest(ResourceRequest&& proposedRequest, IPC::
     LOG(Network, "(WebProcess) WebResourceLoader::willSendRequest to '%s'", proposedRequest.url().string().latin1().data());
     WEBRESOURCELOADER_RELEASE_LOG(WEBRESOURCELOADER_WILLSENDREQUEST);
     
-    if (RefPtr frame = m_coreLoader->frame()) {
```
Ditto about declaring a local variable.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -169,58 +169,57 @@ void WebResourceLoader::didReceiveResponse(ResourceResponse&& response, PrivateR
-            // If m_coreLoader becomes null as a result of the didReceiveResponse callback, we can't use the send function().
-            if (m_coreLoader && m_coreLoader->identifier())
+            // If coreLoader becomes null as a result of the didReceiveResponse callback, we can't use the send function().
+            if (m_coreLoader && protectedCoreLoader()->identifier())
```
Here too, we probably want to declare a local variable:
```
RefPtr coreLoader = m_coreLoader;
```

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -169,58 +169,57 @@ void WebResourceLoader::didReceiveResponse(ResourceResponse&& response, PrivateR
 void WebResourceLoader::didReceiveData(IPC::SharedBufferReference&& data, uint64_t encodedDataLength)
 {
-    LOG(Network, "(WebProcess) WebResourceLoader::didReceiveData of size %lu for '%s'", data.size(), m_coreLoader->url().string().latin1().data());
+    LOG(Network, "(WebProcess) WebResourceLoader::didReceiveData of size %lu for '%s'", data.size(), protectedCoreLoader()->url().string().latin1().data());
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -231,16 +230,16 @@ void WebResourceLoader::didReceiveData(IPC::SharedBufferReference&& data, uint64
 void WebResourceLoader::didFinishResourceLoad(NetworkLoadMetrics&& networkLoadMetrics)
 {
-    LOG(Network, "(WebProcess) WebResourceLoader::didFinishResourceLoad for '%s'", m_coreLoader->url().string().latin1().data());
+    LOG(Network, "(WebProcess) WebResourceLoader::didFinishResourceLoad for '%s'", protectedCoreLoader()->url().string().latin1().data());
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -250,15 +249,15 @@ void WebResourceLoader::didFinishResourceLoad(NetworkLoadMetrics&& networkLoadMe
 {
-    if (RefPtr document = m_coreLoader->frame() ? m_coreLoader->frame()->document() : nullptr) {
-        if (m_coreLoader->options().destination != FetchOptions::Destination::EmptyString || error.isGeneral())
+    if (RefPtr document = protectedCoreLoader()->frame() ? protectedCoreLoader()->frame()->document() : nullptr) {
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -269,19 +268,19 @@ void WebResourceLoader::serviceWorkerDidNotHandle()
 
-    ASSERT(m_coreLoader->options().serviceWorkersMode == ServiceWorkersMode::Only);
-    auto error = internalError(m_coreLoader->request().url());
+    ASSERT(protectedCoreLoader()->options().serviceWorkersMode == ServiceWorkersMode::Only);
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -269,19 +268,19 @@ void WebResourceLoader::serviceWorkerDidNotHandle()
 void WebResourceLoader::didFailResourceLoad(const ResourceError& error)
 {
-    LOG(Network, "(WebProcess) WebResourceLoader::didFailResourceLoad for '%s'", m_coreLoader->url().string().latin1().data());
+    LOG(Network, "(WebProcess) WebResourceLoader::didFailResourceLoad for '%s'", protectedCoreLoader()->url().string().latin1().data());
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -290,71 +289,76 @@ void WebResourceLoader::didFailResourceLoad(const ResourceError& error)
 void WebResourceLoader::didBlockAuthenticationChallenge()
 {
-    LOG(Network, "(WebProcess) WebResourceLoader::didBlockAuthenticationChallenge for '%s'", m_coreLoader->url().string().latin1().data());
+    LOG(Network, "(WebProcess) WebResourceLoader::didBlockAuthenticationChallenge for '%s'", protectedCoreLoader()->url().string().latin1().data());
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -290,71 +289,76 @@ void WebResourceLoader::didFailResourceLoad(const ResourceError& error)
 
 void WebResourceLoader::stopLoadingAfterXFrameOptionsOrContentSecurityPolicyDenied(const ResourceResponse& response)
 {
-    LOG(Network, "(WebProcess) WebResourceLoader::stopLoadingAfterXFrameOptionsOrContentSecurityPolicyDenied for '%s'", m_coreLoader->url().string().latin1().data());
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -290,71 +289,76 @@ void WebResourceLoader::didFailResourceLoad(const ResourceError& error)
 void WebResourceLoader::didReceiveResource(ShareableResource::Handle&& handle)
 {
-    LOG(Network, "(WebProcess) WebResourceLoader::didReceiveResource for '%s'", m_coreLoader->url().string().latin1().data());
+    LOG(Network, "(WebProcess) WebResourceLoader::didReceiveResource for '%s'", protectedCoreLoader()->url().string().latin1().data());
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -290,71 +289,76 @@ void WebResourceLoader::didFailResourceLoad(const ResourceError& error)
 void WebResourceLoader::contentFilterDidBlockLoad(const WebCore::ContentFilterUnblockHandler& unblockHandler, String&& unblockRequestDeniedScript, const ResourceError& error, const URL& blockedPageURL,  WebCore::SubstituteData&& substituteData)
 {
-    if (!m_coreLoader || !m_coreLoader->documentLoader())
+    if (!m_coreLoader || !protectedCoreLoader()->documentLoader())
```
Ditto.

rniwa commented on Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
```diff
@@ -115,39 +116,42 @@ MainFrameMainResource WebResourceLoader::mainFrameMainResource() const
 void WebResourceLoader::didSendData(uint64_t bytesSent, uint64_t totalBytesToBeSent)
 {
-    m_coreLoader->didSendData(bytesSent, totalBytesToBeSent);
+    RefPtr { m_coreLoader }->didSendData(bytesSent, totalBytesToBeSent);
```
We can use `protectedCoreLoader()` here.

lopezlozoya committed:
Implement more smart pointers in WebResourceLoader.h/cpp
https://bugs.webkit.org/show_bug.cgi?id=287211

Reviewed by Ryosuke Niwa.

Addressed the smart pointer static analyzer warnings.

* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebKit/WebProcess/Network/WebResourceLoader.cpp:
* Source/WebKit/WebProcess/Network/WebResourceLoader.h:
(WebKit::WebResourceLoader::messageSenderDestinationID const):
(WebKit::WebResourceLoader::mainFrameMainResource const):
(WebKit::WebResourceLoader::willSendRequest):
(WebKit::WebResourceLoader::didSendData):
(WebKit::WebResourceLoader::didReceiveResponse):
(WebKit::WebResourceLoader::didReceiveData):
(WebKit::WebResourceLoader::didFinishResourceLoad):
(WebKit::WebResourceLoader::didFailServiceWorkerLoad):
(WebKit::WebResourceLoader::serviceWorkerDidNotHandle):
(WebKit::WebResourceLoader::didFailResourceLoad):
(WebKit::WebResourceLoader::didBlockAuthenticationChallenge):
(WebKit::WebResourceLoader::stopLoadingAfterXFrameOptionsOrContentSecurityPolicyDenied):
(WebKit::WebResourceLoader::didReceiveResource):
(WebKit::WebResourceLoader::contentFilterDidBlockLoad):

Canonical link: https://commits.webkit.org/290084@main
